# **Testing ML Algorithms**

## Objectives

* Using accuracy as a readout, to test which 2 algorithms give the best results

## Inputs

* The data file, "US_Accidents_For_ML.csv", which is saved locally in "Data/ML"

## Outputs

* The 2 best performing algorithms which will be carried forward for hyperparamter optimisation

## Summary of Steps

* Load the dataset
* Split data into train and test sets
* Test algorithms with data that has not been transformed
* Test algorithms with data that has been transformed
* Test algorithms with "Start_Lat" and "Start_Lng" columns dropped (28 variables)
* Test algorithms with a reduced set of 20 variables (cut from 30)
* Test algorithms with a reduced set of 20 variables (cut from 30) and transformed numerical variables
* Pick the 2 algorithms, and dataset, that give the best accuracy for processing power 

---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\sonia\\Documents\\VS Studio Projects\\US_Accidents_ML_Project\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\sonia\\Documents\\VS Studio Projects\\US_Accidents_ML_Project'

---

## Required Libraries

In [5]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import QuantileTransformer
from feature_engine.transformation import YeoJohnsonTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier 
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import HistGradientBoostingClassifier
from xgboost import XGBClassifier
from mord import LogisticAT
from sklearn.neural_network import MLPClassifier

---

## Load the Dataset

I will use Pandas to load my dataset into a DataFrame.

In [7]:
df = pd.read_csv("Data/ML/US_Accidents_For_ML.csv")
pd.set_option("display.max_columns", None)
df.head()

Severity  Start_Lat  Start_Lng  Distance(mi) Timezone  Temperature(F)  \
0         2  32.456486 -93.774536         0.501  Central            78.0   
1         2  36.804693 -76.189728         0.253  Eastern            54.0   
2         2  29.895741 -90.090026         1.154  Pacific            40.0   
3         2  32.456459 -93.779709         0.016  Central            62.0   
4         2  26.966433 -82.255414         0.057  Eastern            84.0   

   Humidity(%)  Pressure(in)  Visibility(mi) Wind_Direction  Wind_Speed(mph)  \
0         62.0         29.61            10.0           CALM              0.0   
1         90.0         30.40             7.0           CALM              0.0   
2         58.0         30.28            10.0              N             10.0   
3         75.0         29.80            10.0            SSE              8.0   
4         69.0         29.99            10.0              E             18.0   

   Precipitation(in)  Amenity   Bump  Crossing  Give_Way  Junction  No_Exit  \
0                0.0    False  False     False     False     False    False   
1                0.0    False  False      True     False     False    False   
2                0.0    False  False     False     False      True    False   
3                0.0    False  False     False     False     False    False   
4                0.0    False  False     False     False     False    False   

   Railway  Station   Stop  Traffic_Calming  Traffic_Signal Sunrise_Sunset  \
0    False    False  False            False           False          Night   
1    False    False  False            False            True          Night   
2    False    False  False            False           False            Day   
3    False    False  False            False           False          Night   
4    False    False  False            False           False            Day   

  Clearance_Class Weather_Simplified State_Other  Population    County_Other  \
0       Very Long               Fair          LA      187540           Caddo   
1       Very Long               Fair          VA      459444  Virginia Beach   
2       Very Long             Cloudy          LA      440784       Jefferson   
3       Very Long             Cloudy          LA      187540           Caddo   
4       Very Long             Cloudy          FL      186824           Other   

  Month  
0   Sep  
1   May  
2   Jan  
3   Nov  
4   Sep

---

## Split into Train and Test

First, I will split the data into train (80 %) and test (80 %) sets. I am also encoding the target, "Clearance_Class", for use with the ordinal logistic regression algorithm, "LogisticAT".

In [8]:
mapping = {'Short': 0, 'Moderate': 1, 'Long': 2, "Very Long": 3}
df['Clearance_Class_num'] = df['Clearance_Class'].map(mapping)

X_train, X_test, y_train, y_test = train_test_split(
    df.drop(['Clearance_Class', "Clearance_Class_num"], axis=1),
    df['Clearance_Class_num'],
    test_size=0.2,
    random_state=0
)
print(
    "* Train set:",
    X_train.shape,
    y_train.shape,
    "\n* Test set:",
    X_test.shape,
    y_test.shape,
)

* Train set: (7968, 29) (7968,) 
* Test set: (1992, 29) (1992,)


Now I am ready to compare algorithms.

---

## Comparison of Algorithms - No Transformed Variables

Here, I have kept the variables "Start_Lat" and "Start_Lng", which are treated the same as other numerical data and scaled. Categorical variables are encoded with OneHotEncoder i.e. randomly. 

Steps to be included in the preprocessor step:

- Encoding categorical variables
- Scaling numerical variables

Steps to be included in the pipeline:
- Preprocessor
- Feature selection
- Model selection 

In [9]:
def PipelineOptimization(model, X):
    numeric_cols = X.select_dtypes(include="number").columns.tolist()
    categorical_cols = X.select_dtypes(include="object").columns.tolist()

    numeric_pipeline = Pipeline([
        ("scaler", StandardScaler())
    ])

    categorical_pipeline = Pipeline([
        ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
    ])

    preprocessor = ColumnTransformer(transformers=[
        ("numeric", numeric_pipeline, numeric_cols),
        ("categorical", categorical_pipeline, categorical_cols)
    ])

    steps = [("preprocessor", preprocessor)]

    # Add feature selection if the model supports it
    if hasattr(model, "coef_") or hasattr(model, "feature_importances_"):
        steps.append(("feat_selection", SelectFromModel(model)))

    steps.append(("model", model))

    pipeline = Pipeline(steps)
    return pipeline

These are the algorithms I am testing. They are mostly tree and ensemle tree based, but I also wanted to test a neural network (MLPClassifier) and logistic regression (LogisticAT) based agorithm.

In [ ]:
models_search = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=0),
    "RandomForestClassifier": RandomForestClassifier(random_state=0),
    "GradientBoostingClassifier": GradientBoostingClassifier(random_state=0),
    "ExtraTreesClassifier": ExtraTreesClassifier(random_state=0),
    "AdaBoostClassifier": AdaBoostClassifier(random_state=0),
    "HGrBoostingClassifier": HistGradientBoostingClassifier(random_state=0),
    "Xgb": XGBClassifier(random_state=0), 
    "OrdinalLogistic": LogisticAT(),
    "Mlp": MLPClassifier(random_state=0)
}

At this first stage, I am using default hyperparameter settings.

In [ ]:
params_search = {
    "DecisionTreeClassifier": {},
    "RandomForestClassifier": {},
    "GradientBoostingClassifier": {},
    "ExtraTreesClassifier": {},
    "AdaBoostClassifier": {},
    "HGrBoostingClassifier": {},
    "Xgb": {},
    "OrdinalLogistic": {},
    "Mlp": {}
}

This function uses GridSearchCV to run grid searches for multiple models and hyperparameters, specified above, in a loop, saves their results and then returns a ranked, summary table of their performance.

In [12]:
class HyperparameterOptimizationSearch:

    def __init__(self, models, params):
        self.models = models
        self.params = params
        self.keys = list(models.keys())
        self.grid_searches = {}

    def fit(self, X, y, cv, n_jobs, verbose=1, scoring=None, refit=False):
        for key in self.keys:
            print(f"\nRunning GridSearchCV for {key} \n")
            model_pipeline = PipelineOptimization(self.models[key], X)
            params = self.params[key]
            gs = GridSearchCV(model_pipeline, params, cv=cv, n_jobs=n_jobs, verbose=verbose, scoring=scoring, refit=refit)
            gs.fit(X, y)
            self.grid_searches[key] = gs

    def score_summary(self, sort_by='mean_score'):
        def row(key, scores, params):
            d = {
                'estimator': key,
                'min_score': min(scores),
                'max_score': max(scores),
                'mean_score': np.mean(scores),
                'std_score': np.std(scores),
            }
            return pd.Series({**params, **d})

        rows = []
        for k in self.grid_searches:
            gs = self.grid_searches[k]
            params = gs.cv_results_['params']
            all_scores = gs.cv_results_['mean_test_score']
            for p, s in zip(params, all_scores):
                rows.append(row(k, [s], p))

        df = pd.concat(rows, axis=1).T.sort_values([sort_by], ascending=False)
        columns = ['estimator', 'min_score', 'mean_score', 'max_score', 'std_score']
        columns += [c for c in df.columns if c not in columns]
        return df[columns], self.grid_searches

Now I train the models, creating an instance of HyperparameterOptimizationSearch, which contains the algorithms and hyperparameters to use.

Cross validation (cv) is used to equaluate how well the model generalises. It splits the train set into cv=*n*  parts (folds), and then runs *n* rounds of training and testing, rotating which fold is tested each time. 
Here, I use cv=2, which is OK for big datasets and experimenting as I'm doing here. Usually, cv=5 is the minimum used to gain accurate results. 

In [13]:
search = HyperparameterOptimizationSearch(models=models_search, params=params_search)
search.fit(
    X_train, y_train,
    scoring='accuracy',
    n_jobs=-1,
    cv=2
)


Running GridSearchCV for DecisionTreeClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for RandomForestClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for GradientBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for ExtraTreesClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for AdaBoostClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for HGrBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for xgb 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for OrdinalLogistic 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for mlp 

Fitting 2 folds for each of 1 candidates, totalling 2 fits


In [14]:
grid_search_summary, grid_search_pipelines = search.score_summary(sort_by='mean_score')
grid_search_summary

estimator min_score mean_score max_score std_score
1      RandomForestClassifier  0.583835   0.583835  0.583835       0.0
6                         xgb  0.580949   0.580949  0.580949       0.0
2  GradientBoostingClassifier  0.580572   0.580572  0.580572       0.0
5       HGrBoostingClassifier   0.57756    0.57756   0.57756       0.0
4          AdaBoostClassifier  0.529116   0.529116  0.529116       0.0
3        ExtraTreesClassifier  0.527108   0.527108  0.527108       0.0
0      DecisionTreeClassifier       0.5        0.5       0.5       0.0
8                         mlp  0.496737   0.496737  0.496737       0.0
7             OrdinalLogistic  0.442897   0.442897  0.442897       0.0

The best performing models are RandomForestClassifier (accuracy = 58.4 %) followed by XGBoost (accuracy = 58.1 %).

---

## Comparison of Algorithms - Transformed Variables

Above, I find that the Mlp and OrdinalLogistic models performed worst. These algorithms are more sensitive to skewed data, while tree based algorithms are generally not. Now, I will transform the numeric variables as optimised in the previous notebook "Number_Feature_Transformation", and train the models again to see if this improves accuracy.

I will not be transforming "Start_Lat" and "Start_Lng", but will be scaling them as done above.

Steps to be included in the preprocessor step:

- Transformation of numeric variables (except "Start_Lat" and "Start_Lng")
- Scaling of numeric variables
- Encoding categoric variables

Steps to be included in the pipeline:
- Preprocessor
- Feature selection
- Model selection 

In [15]:
def PipelineOptimization(model, X):
    numeric_yeojohnson = ["Distance(mi)", "Humidity(%)", "Population", "Pressure(in)", "Wind_Speed(mph)"]
    numeric_quantile = ["Temperature(F)", "Visibility(mi)", "Precipitation(in)"]
    numeric_coords = ["Start_Lat", "Start_Lng"]
    categorical_cols = X.select_dtypes(include="object").columns.tolist()

    yeojohnson_pipeline = Pipeline([
        ("yeojohnson", YeoJohnsonTransformer(variables=numeric_yeojohnson)),
        ("scaler", StandardScaler())
    ])

    quantile_pipeline = Pipeline([
        ("quantile", QuantileTransformer(output_distribution="normal")),
        ("scaler", StandardScaler())
    ])

    coords_pipeline = Pipeline([
        ("scaler", StandardScaler())
    ])

    categorical_pipeline = Pipeline([
        ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
    ])

    preprocessor = ColumnTransformer(transformers=[
        ("yeojohnson", yeojohnson_pipeline, numeric_yeojohnson),
        ("quantile", quantile_pipeline, numeric_quantile),
        ("coords", coords_pipeline, numeric_coords),
        ("categorical", categorical_pipeline, categorical_cols)
    ])

    steps = [("preprocessor", preprocessor)]

    # Add feature selection if the model supports it
    if hasattr(model, "coef_") or hasattr(model, "feature_importances_"):
        steps.append(("feat_selection", SelectFromModel(model)))

    steps.append(("model", model))

    pipeline = Pipeline(steps)
    return pipeline

In [ ]:
models_search = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=0),
    "RandomForestClassifier": RandomForestClassifier(random_state=0),
    "GradientBoostingClassifier": GradientBoostingClassifier(random_state=0),
    "ExtraTreesClassifier": ExtraTreesClassifier(random_state=0),
    "AdaBoostClassifier": AdaBoostClassifier(random_state=0),
    "HGrBoostingClassifier": HistGradientBoostingClassifier(random_state=0),
    "Xgb": XGBClassifier(random_state=0), 
    "OrdinalLogistic": LogisticAT(),
    "Mlp": MLPClassifier(random_state=0)
}

In [ ]:
params_search = {
    "DecisionTreeClassifier": {},
    "RandomForestClassifier": {},
    "GradientBoostingClassifier": {},
    "ExtraTreesClassifier": {},
    "AdaBoostClassifier": {},
    "HGrBoostingClassifier": {},
    "Xgb": {},
    "OrdinalLogistic": {},
    "Mlp": {}
}

In [18]:
class HyperparameterOptimizationSearch:

    def __init__(self, models, params):
        self.models = models
        self.params = params
        self.keys = list(models.keys())
        self.grid_searches = {}

    def fit(self, X, y, cv, n_jobs, verbose=1, scoring=None, refit=False):
        for key in self.keys:
            print(f"\nRunning GridSearchCV for {key} \n")
            model_pipeline = PipelineOptimization(self.models[key], X)
            params = self.params[key]
            gs = GridSearchCV(model_pipeline, params, cv=cv, n_jobs=n_jobs, verbose=verbose, scoring=scoring, refit=refit)
            gs.fit(X, y)
            self.grid_searches[key] = gs

    def score_summary(self, sort_by='mean_score'):
        def row(key, scores, params):
            d = {
                'estimator': key,
                'min_score': min(scores),
                'max_score': max(scores),
                'mean_score': np.mean(scores),
                'std_score': np.std(scores),
            }
            return pd.Series({**params, **d})

        rows = []
        for k in self.grid_searches:
            gs = self.grid_searches[k]
            params = gs.cv_results_['params']
            all_scores = gs.cv_results_['mean_test_score']
            for p, s in zip(params, all_scores):
                rows.append(row(k, [s], p))

        df = pd.concat(rows, axis=1).T.sort_values([sort_by], ascending=False)
        columns = ['estimator', 'min_score', 'mean_score', 'max_score', 'std_score']
        columns += [c for c in df.columns if c not in columns]
        return df[columns], self.grid_searches

In [19]:
search = HyperparameterOptimizationSearch(models=models_search, params=params_search)
search.fit(
    X_train, y_train,
    scoring='accuracy',
    n_jobs=-1,
    cv=2
)


Running GridSearchCV for DecisionTreeClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for RandomForestClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for GradientBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for ExtraTreesClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for AdaBoostClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for HGrBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for xgb 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for OrdinalLogistic 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for mlp 

Fitting 2 folds for each of 1 candidates, totalling 2 fits


In [20]:
grid_search_summary, grid_search_pipelines = search.score_summary(sort_by='mean_score')
grid_search_summary

estimator min_score mean_score max_score std_score
1      RandomForestClassifier  0.574297   0.574297  0.574297       0.0
2  GradientBoostingClassifier  0.572917   0.572917  0.572917       0.0
6                         xgb  0.565136   0.565136  0.565136       0.0
5       HGrBoostingClassifier  0.562374   0.562374  0.562374       0.0
3        ExtraTreesClassifier  0.530748   0.530748  0.530748       0.0
4          AdaBoostClassifier  0.529493   0.529493  0.529493       0.0
8                         mlp  0.499874   0.499874  0.499874       0.0
0      DecisionTreeClassifier   0.48883    0.48883   0.48883       0.0
7             OrdinalLogistic  0.467997   0.467997  0.467997       0.0

The best performing models with transformed data are RandomForestClassifier (accuracy = 57.4 %) followed by GradientBoostingClassifier (accuracy = 57.3 %).

Compared with the non-transformed data above, the transformed dataset performed slightly worse overall, with the exception being OrdinalLogistic, whose accuracy went from 44.3 % to 46.8 %.

---

## Effect of Latitude and Longitude data

I am interested in understanding whether the agorithms will perform better without "Start_Lat" and "Start_Lng".  I am going to repeat the training without these variables and using non-transformed data. 

In [23]:
keep_col = [
    "Severity",
    "Distance(mi)",
    "Timezone",
    "Temperature(F)",
    "Humidity(%)",
    "Pressure(in)",
    "Visibility(mi)",
    "Wind_Direction",
    "Wind_Speed(mph)",
    "Precipitation(in)",
    "Amenity",
    "Bump",
    "Crossing",
    "Give_Way",
    "Junction",
    "No_Exit",
    "Railway",
    "Station",
    "Stop",
    "Traffic_Calming",
    "Traffic_Signal",
    "Sunrise_Sunset",
    "Clearance_Class",
    "Weather_Simplified",
    "State_Other",
    "Population",
    "County_Other",
    "Month"
]

df_keep = df[keep_col].copy()
df_keep.head()

Severity  Distance(mi) Timezone  Temperature(F)  Humidity(%)  Pressure(in)  \
0         2         0.501  Central            78.0         62.0         29.61   
1         2         0.253  Eastern            54.0         90.0         30.40   
2         2         1.154  Pacific            40.0         58.0         30.28   
3         2         0.016  Central            62.0         75.0         29.80   
4         2         0.057  Eastern            84.0         69.0         29.99   

   Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  Amenity  \
0            10.0           CALM              0.0                0.0    False   
1             7.0           CALM              0.0                0.0    False   
2            10.0              N             10.0                0.0    False   
3            10.0            SSE              8.0                0.0    False   
4            10.0              E             18.0                0.0    False   

    Bump  Crossing  Give_Way  Junction  No_Exit  Railway  Station   Stop  \
0  False     False     False     False    False    False    False  False   
1  False      True     False     False    False    False    False  False   
2  False     False     False      True    False    False    False  False   
3  False     False     False     False    False    False    False  False   
4  False     False     False     False    False    False    False  False   

   Traffic_Calming  Traffic_Signal Sunrise_Sunset Clearance_Class  \
0            False           False          Night       Very Long   
1            False            True          Night       Very Long   
2            False           False            Day       Very Long   
3            False           False          Night       Very Long   
4            False           False            Day       Very Long   

  Weather_Simplified State_Other  Population    County_Other Month  
0               Fair          LA      187540           Caddo   Sep  
1               Fair          VA      459444  Virginia Beach   May  
2             Cloudy          LA      440784       Jefferson   Jan  
3             Cloudy          LA      187540           Caddo   Nov  
4             Cloudy          FL      186824           Other   Sep

In [24]:
mapping = {'Short': 0, 'Moderate': 1, 'Long': 2, "Very Long": 3}
df_keep['Clearance_Class_num'] = df_keep['Clearance_Class'].map(mapping)

X_train, X_test, y_train, y_test = train_test_split(
    df_keep.drop(['Clearance_Class', "Clearance_Class_num"], axis=1),
    df_keep['Clearance_Class_num'],
    test_size=0.2,
    random_state=0
)
print(
    "* Train set:",
    X_train.shape,
    y_train.shape,
    "\n* Test set:",
    X_test.shape,
    y_test.shape,
)

* Train set: (7968, 27) (7968,) 
* Test set: (1992, 27) (1992,)


In [30]:
def PipelineOptimization(model, X):
    numeric_cols = X.select_dtypes(include="number").columns.tolist()
    categorical_cols = X.select_dtypes(include="object").columns.tolist()

    numeric_pipeline = Pipeline([
        ("scaler", StandardScaler())
    ])

    categorical_pipeline = Pipeline([
        ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
    ])

    preprocessor = ColumnTransformer(transformers=[
        ("numeric", numeric_pipeline, numeric_cols),
        ("categorical", categorical_pipeline, categorical_cols)
    ])

    steps = [("preprocessor", preprocessor)]

    # Add feature selection if the model supports it
    if hasattr(model, "coef_") or hasattr(model, "feature_importances_"):
        steps.append(("feat_selection", SelectFromModel(model)))

    steps.append(("model", model))

    pipeline = Pipeline(steps)
    return pipeline

In [ ]:
models_search = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=0),
    "RandomForestClassifier": RandomForestClassifier(random_state=0),
    "GradientBoostingClassifier": GradientBoostingClassifier(random_state=0),
    "ExtraTreesClassifier": ExtraTreesClassifier(random_state=0),
    "AdaBoostClassifier": AdaBoostClassifier(random_state=0),
    "HGrBoostingClassifier": HistGradientBoostingClassifier(random_state=0),
    "Xgb": XGBClassifier(random_state=0), 
    "OrdinalLogistic": LogisticAT(),
    "Mlp": MLPClassifier(random_state=0)
}

In [ ]:
params_search = {
    "DecisionTreeClassifier": {},
    "RandomForestClassifier": {},
    "GradientBoostingClassifier": {},
    "ExtraTreesClassifier": {},
    "AdaBoostClassifier": {},
    "HGrBoostingClassifier": {},
    "Xgb": {},
    "OrdinalLogistic": {},
    "Mlp": {}
}

In [28]:
class HyperparameterOptimizationSearch:

    def __init__(self, models, params):
        self.models = models
        self.params = params
        self.keys = list(models.keys())
        self.grid_searches = {}

    def fit(self, X, y, cv, n_jobs, verbose=1, scoring=None, refit=False):
        for key in self.keys:
            print(f"\nRunning GridSearchCV for {key} \n")
            model_pipeline = PipelineOptimization(self.models[key], X)
            params = self.params[key]
            gs = GridSearchCV(model_pipeline, params, cv=cv, n_jobs=n_jobs, verbose=verbose, scoring=scoring, refit=refit)
            gs.fit(X, y)
            self.grid_searches[key] = gs

    def score_summary(self, sort_by='mean_score'):
        def row(key, scores, params):
            d = {
                'estimator': key,
                'min_score': min(scores),
                'max_score': max(scores),
                'mean_score': np.mean(scores),
                'std_score': np.std(scores),
            }
            return pd.Series({**params, **d})

        rows = []
        for k in self.grid_searches:
            gs = self.grid_searches[k]
            params = gs.cv_results_['params']
            all_scores = gs.cv_results_['mean_test_score']
            for p, s in zip(params, all_scores):
                rows.append(row(k, [s], p))

        df = pd.concat(rows, axis=1).T.sort_values([sort_by], ascending=False)
        columns = ['estimator', 'min_score', 'mean_score', 'max_score', 'std_score']
        columns += [c for c in df.columns if c not in columns]
        return df[columns], self.grid_searches

In [31]:
search = HyperparameterOptimizationSearch(models=models_search, params=params_search)
search.fit(
    X_train, y_train,
    scoring='accuracy',
    n_jobs=-1,
    cv=2
)


Running GridSearchCV for DecisionTreeClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for RandomForestClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for GradientBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for ExtraTreesClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for AdaBoostClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for HGrBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for xgb 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for OrdinalLogistic 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for mlp 

Fitting 2 folds for each of 1 candidates, totalling 2 fits


In [32]:
grid_search_summary, grid_search_pipelines = search.score_summary(sort_by='mean_score')
grid_search_summary

estimator min_score mean_score max_score std_score
2  GradientBoostingClassifier  0.585216   0.585216  0.585216       0.0
1      RandomForestClassifier  0.584839   0.584839  0.584839       0.0
6                         xgb  0.579066   0.579066  0.579066       0.0
5       HGrBoostingClassifier  0.572415   0.572415  0.572415       0.0
4          AdaBoostClassifier   0.53012    0.53012   0.53012       0.0
3        ExtraTreesClassifier  0.527108   0.527108  0.527108       0.0
8                         mlp  0.497741   0.497741  0.497741       0.0
0      DecisionTreeClassifier  0.489081   0.489081  0.489081       0.0
7             OrdinalLogistic  0.444528   0.444528  0.444528       0.0

The best performing models are GradientBoostingClassifier (accuracy = 58.5 %) followed by RandomForestClassifier (accuracy = 58.5 %).

This version of the non-transformed dataset without the variables "Start_Lat" and "Start_Lng", has a very marginally increased accuracy by ~0.1 %; the best performing so far, and also likely saves on processing power.

---

## Effect of Dropping Non-Significant Variables 

As dropping "Start_Lat" and "Start_Lng" has a small favourable outcome, I will try dropping more variables before training the models again. I am cutting the boolean variables "Amenity, "Bump", "Crossing", "Give-Way", "Junction", "No_Exit", "Railway" and "Traffic_Calming", which did not show any statistically significant differences between clearance classes or clearance time.

In [52]:
keep_col = [
    "Severity",
    "Distance(mi)",
    "Timezone",
    "Temperature(F)",
    "Humidity(%)",
    "Pressure(in)",
    "Visibility(mi)",
    "Wind_Direction",
    "Wind_Speed(mph)",
    "Precipitation(in)",
    "Station",
    "Stop",
    "Traffic_Signal",
    "Sunrise_Sunset",
    "Clearance_Class",
    "Weather_Simplified",
    "State_Other",
    "Population",
    "County_Other",
    "Month"
]

df_keep = df[keep_col].copy()
df_keep.head()

Severity  Distance(mi) Timezone  Temperature(F)  Humidity(%)  Pressure(in)  \
0         2         0.501  Central            78.0         62.0         29.61   
1         2         0.253  Eastern            54.0         90.0         30.40   
2         2         1.154  Pacific            40.0         58.0         30.28   
3         2         0.016  Central            62.0         75.0         29.80   
4         2         0.057  Eastern            84.0         69.0         29.99   

   Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  Station  \
0            10.0           CALM              0.0                0.0    False   
1             7.0           CALM              0.0                0.0    False   
2            10.0              N             10.0                0.0    False   
3            10.0            SSE              8.0                0.0    False   
4            10.0              E             18.0                0.0    False   

    Stop  Traffic_Signal Sunrise_Sunset Clearance_Class Weather_Simplified  \
0  False           False          Night       Very Long               Fair   
1  False            True          Night       Very Long               Fair   
2  False           False            Day       Very Long             Cloudy   
3  False           False          Night       Very Long             Cloudy   
4  False           False            Day       Very Long             Cloudy   

  State_Other  Population    County_Other Month  
0          LA      187540           Caddo   Sep  
1          VA      459444  Virginia Beach   May  
2          LA      440784       Jefferson   Jan  
3          LA      187540           Caddo   Nov  
4          FL      186824           Other   Sep

In [53]:
mapping = {'Short': 0, 'Moderate': 1, 'Long': 2, "Very Long": 3}
df_keep['Clearance_Class_num'] = df_keep['Clearance_Class'].map(mapping)

X_train, X_test, y_train, y_test = train_test_split(
    df_keep.drop(['Clearance_Class', "Clearance_Class_num"], axis=1),
    df_keep['Clearance_Class_num'],
    test_size=0.2,
    random_state=0
)
print(
    "* Train set:",
    X_train.shape,
    y_train.shape,
    "\n* Test set:",
    X_test.shape,
    y_test.shape,
)

* Train set: (7968, 19) (7968,) 
* Test set: (1992, 19) (1992,)


In [54]:
def PipelineOptimization(model, X):
    numeric_cols = X.select_dtypes(include="number").columns.tolist()
    categorical_cols = X.select_dtypes(include="object").columns.tolist()

    numeric_pipeline = Pipeline([
        ("scaler", StandardScaler())
    ])

    categorical_pipeline = Pipeline([
        ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
    ])

    preprocessor = ColumnTransformer(transformers=[
        ("numeric", numeric_pipeline, numeric_cols),
        ("categorical", categorical_pipeline, categorical_cols)
    ])

    steps = [("preprocessor", preprocessor)]

    # Add feature selection if the model supports it
    if hasattr(model, "coef_") or hasattr(model, "feature_importances_"):
        steps.append(("feat_selection", SelectFromModel(model)))

    steps.append(("model", model))

    pipeline = Pipeline(steps)
    return pipeline

In [55]:
models_search = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=0),
    "RandomForestClassifier": RandomForestClassifier(random_state=0),
    "GradientBoostingClassifier": GradientBoostingClassifier(random_state=0),
    "ExtraTreesClassifier": ExtraTreesClassifier(random_state=0),
    "AdaBoostClassifier": AdaBoostClassifier(random_state=0),
    "HGrBoostingClassifier": HistGradientBoostingClassifier(random_state=0),
    "Xgb": XGBClassifier(random_state=0), 
    "OrdinalLogistic": LogisticAT(),
    "Mlp": MLPClassifier(random_state=0)
}

In [56]:
params_search = {
    "DecisionTreeClassifier": {},
    "RandomForestClassifier": {},
    "GradientBoostingClassifier": {},
    "ExtraTreesClassifier": {},
    "AdaBoostClassifier": {},
    "HGrBoostingClassifier": {},
    "Xgb": {},
    "OrdinalLogistic": {},
    "Mlp": {}
}

In [57]:
class HyperparameterOptimizationSearch:

    def __init__(self, models, params):
        self.models = models
        self.params = params
        self.keys = list(models.keys())
        self.grid_searches = {}

    def fit(self, X, y, cv, n_jobs, verbose=1, scoring=None, refit=False):
        for key in self.keys:
            print(f"\nRunning GridSearchCV for {key} \n")
            model_pipeline = PipelineOptimization(self.models[key], X)
            params = self.params[key]
            gs = GridSearchCV(model_pipeline, params, cv=cv, n_jobs=n_jobs, verbose=verbose, scoring=scoring, refit=refit)
            gs.fit(X, y)
            self.grid_searches[key] = gs

    def score_summary(self, sort_by='mean_score'):
        def row(key, scores, params):
            d = {
                'estimator': key,
                'min_score': min(scores),
                'max_score': max(scores),
                'mean_score': np.mean(scores),
                'std_score': np.std(scores),
            }
            return pd.Series({**params, **d})

        rows = []
        for k in self.grid_searches:
            gs = self.grid_searches[k]
            params = gs.cv_results_['params']
            all_scores = gs.cv_results_['mean_test_score']
            for p, s in zip(params, all_scores):
                rows.append(row(k, [s], p))

        df = pd.concat(rows, axis=1).T.sort_values([sort_by], ascending=False)
        columns = ['estimator', 'min_score', 'mean_score', 'max_score', 'std_score']
        columns += [c for c in df.columns if c not in columns]
        return df[columns], self.grid_searches

In [ ]:
search = HyperparameterOptimizationSearch(models=models_search, params=params_search)
search.fit(
    X_train, y_train,
    scoring='accuracy',
    n_jobs=-1,
    cv=2
)


Running GridSearchCV for DecisionTreeClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for RandomForestClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for GradientBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for ExtraTreesClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for AdaBoostClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for HGrBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for Xgb 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for OrdinalLogistic 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for Mlp 

Fitting 2 folds for each of 1 candidates, totalling 2 fits


In [61]:
grid_search_summary, grid_search_pipelines = search.score_summary(sort_by='mean_score')
grid_search_summary

estimator min_score mean_score max_score std_score
1      RandomForestClassifier   0.58988    0.58988   0.58988       0.0
2  GradientBoostingClassifier   0.58805    0.58805   0.58805       0.0
6                         Xgb  0.583209   0.583209  0.583209       0.0
5       HGrBoostingClassifier  0.575532   0.575532  0.575532       0.0
3        ExtraTreesClassifier  0.538017   0.538017  0.538017       0.0
4          AdaBoostClassifier  0.526799   0.526799  0.526799       0.0
8                         Mlp  0.504527   0.504527  0.504527       0.0
0      DecisionTreeClassifier  0.496082   0.496082  0.496082       0.0
7             OrdinalLogistic  0.459157   0.459157  0.459157       0.0

The reduction to 20 variables from 28, hasn't resulted in any loss in accuracy. This is a favourable outcome due to the decreased processing power required to get the same result.

---

## Reduced Dataset with Transformation

For completeness, I am going to test the reduced dataset together with the transformation of numerical variables.

In [41]:
def PipelineOptimization(model, X):
    numeric_yeojohnson = ["Distance(mi)", "Humidity(%)", "Population", "Pressure(in)", "Wind_Speed(mph)"]
    numeric_quantile = ["Temperature(F)", "Visibility(mi)", "Precipitation(in)"]
    categorical_cols = X.select_dtypes(include="object").columns.tolist()

    yeojohnson_pipeline = Pipeline([
        ("yeojohnson", YeoJohnsonTransformer(variables=numeric_yeojohnson)),
        ("scaler", StandardScaler())
    ])

    quantile_pipeline = Pipeline([
        ("quantile", QuantileTransformer(output_distribution="normal")),
        ("scaler", StandardScaler())
    ])

    categorical_pipeline = Pipeline([
        ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
    ])

    preprocessor = ColumnTransformer(transformers=[
        ("yeojohnson", yeojohnson_pipeline, numeric_yeojohnson),
        ("quantile", quantile_pipeline, numeric_quantile),
        ("categorical", categorical_pipeline, categorical_cols)
    ])

    steps = [("preprocessor", preprocessor)]

    # Add feature selection if the model supports it
    if hasattr(model, "coef_") or hasattr(model, "feature_importances_"):
        steps.append(("feat_selection", SelectFromModel(model)))

    steps.append(("model", model))

    pipeline = Pipeline(steps)
    return pipeline

In [ ]:
models_search = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=0),
    "RandomForestClassifier": RandomForestClassifier(random_state=0),
    "GradientBoostingClassifier": GradientBoostingClassifier(random_state=0),
    "ExtraTreesClassifier": ExtraTreesClassifier(random_state=0),
    "AdaBoostClassifier": AdaBoostClassifier(random_state=0),
    "HGrBoostingClassifier": HistGradientBoostingClassifier(random_state=0),
    "Xgb": XGBClassifier(random_state=0), 
    "OrdinalLogistic": LogisticAT(),
    "Mlp": MLPClassifier(random_state=0)
}

In [ ]:
params_search = {
    "DecisionTreeClassifier": {},
    "RandomForestClassifier": {},
    "GradientBoostingClassifier": {},
    "ExtraTreesClassifier": {},
    "AdaBoostClassifier": {},
    "HGrBoostingClassifier": {},
    "Xgb": {},
    "OrdinalLogistic": {},
    "Mlp": {}
}

In [44]:
class HyperparameterOptimizationSearch:

    def __init__(self, models, params):
        self.models = models
        self.params = params
        self.keys = list(models.keys())
        self.grid_searches = {}

    def fit(self, X, y, cv, n_jobs, verbose=1, scoring=None, refit=False):
        for key in self.keys:
            print(f"\nRunning GridSearchCV for {key} \n")
            model_pipeline = PipelineOptimization(self.models[key], X)
            params = self.params[key]
            gs = GridSearchCV(model_pipeline, params, cv=cv, n_jobs=n_jobs, verbose=verbose, scoring=scoring, refit=refit)
            gs.fit(X, y)
            self.grid_searches[key] = gs

    def score_summary(self, sort_by='mean_score'):
        def row(key, scores, params):
            d = {
                'estimator': key,
                'min_score': min(scores),
                'max_score': max(scores),
                'mean_score': np.mean(scores),
                'std_score': np.std(scores),
            }
            return pd.Series({**params, **d})

        rows = []
        for k in self.grid_searches:
            gs = self.grid_searches[k]
            params = gs.cv_results_['params']
            all_scores = gs.cv_results_['mean_test_score']
            for p, s in zip(params, all_scores):
                rows.append(row(k, [s], p))

        df = pd.concat(rows, axis=1).T.sort_values([sort_by], ascending=False)
        columns = ['estimator', 'min_score', 'mean_score', 'max_score', 'std_score']
        columns += [c for c in df.columns if c not in columns]
        return df[columns], self.grid_searches

In [45]:
search = HyperparameterOptimizationSearch(models=models_search, params=params_search)
search.fit(
    X_train, y_train,
    scoring='accuracy',
    n_jobs=-1,
    cv=2
)


Running GridSearchCV for DecisionTreeClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for RandomForestClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for GradientBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for ExtraTreesClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for AdaBoostClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for HGrBoostingClassifier 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for xgb 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for OrdinalLogistic 

Fitting 2 folds for each of 1 candidates, totalling 2 fits

Running GridSearchCV for mlp 

Fitting 2 folds for each of 1 candidates, totalling 2 fits


In [46]:
grid_search_summary, grid_search_pipelines = search.score_summary(sort_by='mean_score')
grid_search_summary

estimator min_score mean_score max_score std_score
2  GradientBoostingClassifier  0.574297   0.574297  0.574297       0.0
1      RandomForestClassifier  0.573419   0.573419  0.573419       0.0
6                         xgb  0.568148   0.568148  0.568148       0.0
5       HGrBoostingClassifier  0.556476   0.556476  0.556476       0.0
3        ExtraTreesClassifier  0.540412   0.540412  0.540412       0.0
4          AdaBoostClassifier  0.533886   0.533886  0.533886       0.0
8                         mlp  0.493097   0.493097  0.493097       0.0
0      DecisionTreeClassifier  0.491842   0.491842  0.491842       0.0
7             OrdinalLogistic  0.467746   0.467746  0.467746       0.0

Once again, transforming numerical data has not resulted in the best performance overall. However, about half the algorithms, including Mlp, OrdinalLogistic and ExtraTreesClassifier, have shown improved performance using transformed data.

---

## Conclusion and Next Steps

* The 2 best model performances, also taking into account data processing, was found by using a reduced dataset of 20 variables, and the GradientBoostingClassifier and RandomForestClassifier algorithms
* Overall, transformation of numerical data did not improve performance
* The next step is to tune hyperparameters for GradientBoostingClassifier and RandomForestClassifier, with the reduced dataset